### DISCUSSION TITLE
Postgres function returning an array, how to access an item?

### DISCUSSION THREAD

**Discussion by pythonweb2** [sqlalchemy/sqlalchemy#9689] [Usage Questions] [answered]
Here is a simple working example:

```SQL
CREATE FUNCTION public.my_array()
    RETURNS integer[]
    LANGUAGE 'sql'
    
AS $BODY$
SELECT ARRAY[1, 2, 3];
$BODY$;

ALTER FUNCTION public.my_array()
    OWNER TO postgres;
```

Using a statement in SQLAlchemy like this will compile (the cast is needed to make it work).

```python
import sqlalchemy as sa
from sqlalchemy.dialects.postgresql import ARRAY

stmt = sa.select(sa.cast(sa.func.my_array(), ARRAY(sa.Integer, zero_indexes=True))[0])
c = stmt.compile(dialect=postgresql.dialect())
print(c)
print(c.params)
```

And generates this SQL:

```SQL
SELECT CAST(my_array() AS INTEGER[])[%(param_1)s] AS anon_1
{'param_1': 1}
```

However, this will throw a syntax error if run. For some reason (see [this question on stack overflow](https://stackoverflow.com/questions/11831555/get-nth-element-of-an-array-that-returns-from-string-to-array-function)) you need an extra set of parentheses around the function if you want to access items in the array with the square bracket syntax.

For example, if I use this SQL, then it works OK.

```sql
SELECT (CAST(my_array() AS INTEGER[]))[1] AS anon_1
```

Is there a way to add the extra parentheses around the function?

**Comment by CaselIT** [comment, marked as answer]
Hi,

there is no need to cast, just pass the type to the function
`sa.func.my_array(type_=ARRAY(sa.Integer, zero_indexes=True))[0]`

**Reply by pythonweb2** [reply]
Nice! I didn't know that you could specify the return type of a function. Thanks!

**Reply by CaselIT** [reply]
as an alternative if you just need to tell sqlalchemy that something is of a certain type, but you don't want to use cast on the db, you can use [`type_coerce`](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.type_coerce)
